EFCache.Redis/RedisCache.cs

```diff
@@ -109,137 +246,211 @@ public bool GetItem(string key, out object value)
+		/// </summary>
+		/// <param name="key">The text of the SQL query</param>
+		/// <param name="value">The results of the query (byte array)</param>
+		/// <param name="dependentEntitySets">The entity sets (tables) used by the query</param>
+		/// <param name="slidingExpiration">The relative time for expiration</param>
+		/// <param name="absoluteExpiration">The absolute expiration time</param>
 		public void PutItem(string key, object value, IEnumerable<string> dependentEntitySets, TimeSpan slidingExpiration,
 			DateTimeOffset absoluteExpiration)
 		{
 			key.GuardAgainstNullOrEmpty(nameof(key));
-			// ReSharper disable once PossibleMultipleEnumeration - the guard clause should not enumerate, its just checking the reference is not null
-			dependentEntitySets.GuardAgainstNull(nameof(dependentEntitySets));
+			var entitySets = dependentEntitySets as string[] ?? dependentEntitySets.ToArray();
+			entitySets.GuardAgainstNull(nameof(dependentEntitySets));
 
 			var database = Connection.GetDatabase();
 			var hashedKey = HashKey(key);
-			// ReSharper disable once PossibleMultipleEnumeration - the guard clause should not enumerate, its just checking the reference is not null
-			var entitySets = dependentEntitySets.ToArray();
+
+			var keys = new List<RedisKey> {hashedKey};
+			keys.AddRange(entitySets.Select(AddCacheQualifier));
 
 			try
 			{
-				var transaction = database.CreateTransaction();
-				foreach (var entitySet in entitySets)
-				{
-					transaction.SetAddAsync(AddCacheQualifier(entitySet), hashedKey);
-				}
-				transaction.ObjectSetAsync(hashedKey, new CacheEntry(value, entitySets, slidingExpiration, absoluteExpiration));
-				transaction.Execute();
+				var serializedValue = StackExchangeRedisExtensions.Serialize(new CacheEntry(value, entitySets.ToArray(), slidingExpiration,
+					absoluteExpiration));
+				database.ScriptEvaluate(_putItemScriptSha1, keys.ToArray(), new RedisValue[]{serializedValue});
 			}
 			catch (Exception e)
 			{
 				OnCachingFailed(e);
 			}
 		}
 
+		/// <summary>
+		/// Since a sql database write is about to happen, remove all cached sql query results referencing the provided
+		/// entity sets (tables). Then remove the actual entity set values.
+		/// Use a Lua script to ensure that this operation happens atomically.
+		/// </summary>
+		/// <param name="entitySets">The entity sets used by the sql insert, update, delete being issued</param>
 		public void InvalidateSets(IEnumerable<string> entitySets)
 		{
-			// ReSharper disable once PossibleMultipleEnumeration - the guard clause should not enumerate, its just checking the reference is not null
+			// ReSharper disable PossibleMultipleEnumeration
 			entitySets.GuardAgainstNull(nameof(entitySets));
 
 			var database = Connection.GetDatabase();
+			var setKeys = entitySets.Select(AddCacheQualifier);
 
-			try
+			//TODO: Remove this!!
```

Comment:
@jkdufair are we leaving this testing code in for now?

Reply:
I was thinking so earlier but now that you mention it, I think not. I'll just add it in for testing.